### PR TITLE
Mutable properties

### LINKED ISSUE #198: Add Mutability to Properties
Currently only the `issuer` can set a `Property` & if a NFT is `locked` then the `Property` cannot be changed or removed. Based on the spec [setproperty.md](https://github.com/rmrk-team/rmrk-spec/blob/master/standards/rmrk2.0.0/interactions/setproperty.md), the user should be able to customize these properties if they are mutable. We have also run into this problem with flexibility for a project on Phala utilizing RMRK NFTs.

If we take a look at the logic here, the only way to update a `Property` for a NFT is if the Collection `issuer` is the `sender` AND the `issuer` owns the NFT.
https://github.com/rmrk-team/rmrk-substrate/blob/a3abdac27660099baeb50b3db91fadbf66b7e7ea/pallets/rmrk-core/src/functions.rs#L48-L69

### PR DESCRIPTION
**Pull request by arrudagates**
This PR implements the ability to allow the current owner of an NFT to mutate properties that were set as mutable by the collection issuer prior to being sent to the new owner. Closes #198 

Please do let me know if this implementation should be done differently, I just wanted to give a head start to this feature and am flexible to refactoring this PR entirely.

Regarding the arguments to the set_property call including the mutable field, this is handled by ignoring the value of that field if the user is not the collection issuer (which means they can't modify the mutability of the property), but I believe it would be a lot cleaner if we broke it down into 2 distinct calls: set_property and set_property_mutability, the former would be reverted to how it's arguments were previously, and the latter would be responsible for changing the value of the mutable field by the collection issuer (if they're also the current owner of the NFT).

### CODE REVIEW

File: pallets/rmrk-core/src/tests.rs [hunk 1]
```diff
@@ -1814,32 +1821,117 @@ fn set_property_works() {
...
+		assert_ok!(RMRKCore::nft_send(ALICE, 0, 0, AccountIdOrCollectionNftTuple::AccountId(BOB)));
+		// BOB owns the NFT, so he should be able to change the value, but not the mutability
+		assert_ok!(RMRKCore::set_property(
+			Origin::signed(BOB),
+			0,
+			Some(0),
+			key.clone(),
+			value_immutable.clone()
+		));
+		eprintln!("last event");
```
Comment (Szegoo) [inline pallets/rmrk-core/src/tests.rs:1912]: I am not sure I understand why would we print here. And why to stderr? Did you leave this by accident?

**Comment by Szegoo** [comment]
> I believe it would be a lot cleaner if we broke it down into 2 distinct calls: set_property and set_property_mutability

Yes, this would make sense, but I don't think it is bad the way it is now. I am interested to hear the opinions of others here.

File: pallets/rmrk-core/src/tests.rs (hunk 1 above)
Comment (arrudagates) [reply pallets/rmrk-core/src/tests.rs]: Yes, I left that in by accident, oops...

**Review by Szegoo — APPROVED** [review]
LGTM; The integration test will also need to be updated. But that should be probably done after we decide whether we want to have two separate calls, `set_property` and `set_property_mutability`

File: pallets/rmrk-core/src/tests.rs [hunk 2]
```diff
@@ -1814,32 +1821,116 @@ fn set_property_works() {
...
+			value: PropertyValue { mutable: true, value: value_immutable.value.clone() },
+		}));
+		// Property value now be changed
+		assert_eq!(
+			RMRKCore::properties((0, Some(0), key.clone())).unwrap(),
+			PropertyValue { mutable: true, value: value_immutable.value.clone() }
+		);
+	});
+}
+
```
Comment (ilionic) [inline pallets/rmrk-core/src/tests.rs:1926]: Would be good to add extra test for error handling ( `NoPermission` ). Ie when BOB tries to mutate immutable prop.

**Review by ilionic — COMMENTED** [review]
Thank you! Works well. 
I see how splitting `set_property` and `set_property_mutability` would make things cleaner in this case, but need double check few other implementations. Will get back

File: pallets/rmrk-core/src/tests.rs (hunk 2 above)
Comment (arrudagates) [reply pallets/rmrk-core/src/tests.rs:1963]: Added the test case.

**Review by ilionic — APPROVED** [review]
LGTM, thanks! 
Not against splitting `set_property` and `set_property_mutability` if it makes code cleaner. But it seems like might require storage migration, so for @HashWarlock to chime in.
Integration tests could be addressed separately.

**Comment by HashWarlock** [comment]
What would migration for storage look like? This will have a big impact for our PhalaWorld & Stakepool V2 implementation, so we need to prepare for this. 

We should also take a look at how `Attributes` (Uniques v2 NFTs pallet) handles the ability to edit an `Attribute` for an `Item` since the future RMRK pallet will be ported to be tightly coupled with the Uniques v2 Pallet. We can see the implementation here https://github.com/paritytech/substrate/blob/master/frame/nfts/src/features/attributes.rs#L22 with the types settings here https://github.com/paritytech/substrate/blob/248fdf0d4b5e3758cfdadb283b5eca5f0731e466/frame/nfts/src/types.rs#L360

@h4x3rotab may have some input as well on how to best implement this for future compatibility.

**Comment by arrudagates** [comment]
It is possible to separate property mutability into it's own new storage,
which we can then just use as ValueQuery where default is false.

**Pull request closed without merging**